astropy/io/fits/tests/test_table.py

```diff
@@ -3035,6 +3035,32 @@ def test_tolist_from_file(self):
             assert hdu.data.tolist() == [[[45, 56], [11, 3]], [[11, 12, 13], [12, 4]]]
             assert hdu.data['var'].tolist() == [[45, 56], [11, 12, 13]]
 
+    def test_heapsize_P_limit(self):
+        """
+        Regression test for https://github.com/astropy/astropy/issues/10812
+
+        Check if the error is raised when the heap size is bigger than what can be
+        indexed with a 32 bit signed int.
+        """
+
+        # a matrix with variable length array elements is created
+        nrows = 30000
+        matrix = np.zeros(nrows, dtype=np.object_)
+        for i in range(0, nrows):
+            matrix[i] = np.arange(0., float(i+1))
+
+        col = fits.Column(name='MATRIX', format='PD('+str(nrows)+')',
+                          unit='', array=matrix)
+
+        cols = fits.ColDefs([col])
+        t = fits.BinTableHDU.from_columns(cols)
+        t.name = 'MATRIX'
+
+        with pytest.raises(ValueError) as err:
+            t.writeto(self.temp('matrix.fits'))
```

Comment:
I have a test failure I can't understand:
```
E           AssertionError: File(s) not closed:
E             /tmp/fits-test-mh3h7gm3/matrix.fits
```
How should I close this file?
Thanks for any help....

Reply:
Hmm, not quite sure - looks like upon error the file does not get closed, and I do not see another example where an error path is checked. I also wonder if this means the full 2 GB is written out before the error is hit, which would be non-ideal. I fear I may have to leave this to @saimn.

Reply:
But the test that is failing is another one right ?

Reply:
Hmm actually it seems he file is not close in `HDUList.writeto` is there is an exception in the loop on HDUs ...

Reply:
Right.... I'll look into writeto...